internal/controller/environment/controller_finalize.go

```diff
@@ -0,0 +1,172 @@
+	"github.com/openchoreo/openchoreo/internal/labels"
+)
+
+const (
+	// EnvCleanupFinalizer is the finalizer that is used to clean up the environment.
+	EnvCleanupFinalizer = "core.choreo.dev/env-cleanup"
+)
+
+// ensureFinalizer ensures that the finalizer is added to the environment.
+// The first return value indicates whether the finalizer was added to the environment.
+func (r *Reconciler) ensureFinalizer(ctx context.Context, environment *choreov1.Environment) (bool, error) {
+	if !environment.DeletionTimestamp.IsZero() {
+		return false, nil
+	}
+
+	if controllerutil.AddFinalizer(environment, EnvCleanupFinalizer) {
+		return true, r.Update(ctx, environment)
+	}
+
+	return false, nil
+}
+
+// finalize cleans up the data plane resources associated with the environment.
+func (r *Reconciler) finalize(ctx context.Context, old, environment *choreov1.Environment) (ctrl.Result, error) {
+	logger := log.FromContext(ctx).WithValues("environment", environment.Name)
+	if !controllerutil.ContainsFinalizer(environment, EnvCleanupFinalizer) {
+		// Nothing to do if the finalizer is not present
+		return ctrl.Result{}, nil
+	}
+
+	// Mark the environment condition as finalizing and return so that the environment will indicate that it is being finalized.
+	// The actual finalization will be done in the next reconcile loop triggered by the status update.
+	if meta.SetStatusCondition(&environment.Status.Conditions, NewEnvironmentFinalizingCondition(environment.Generation)) {
+		if err := controller.UpdateStatusConditions(ctx, r.Client, old, environment); err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	// Cleaning up the environment.
+	// This assumes that, user already removed the environment from the deployment pipelines.
+
+	// Delete all the deployments associated to the environment.
+	isPending, err := r.cleanupDeployments(ctx, environment)
+
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if isPending {
+		// the next reconcile will be triggered after the pending endpoint/s deleted
+		return ctrl.Result{}, nil
+	}
+
+	// Get the deployment context and delete the data plane resources
+	envCtx, err := r.makeEnvironmentContext(ctx, environment)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to make environment context for finalization: %w", err)
+	}
+
+	resourceHandlers := r.makeExternalResourceHandlers()
+	pendingDpResourcesDeletion := false
+
+	for _, resourceHandler := range resourceHandlers {
+		if err := resourceHandler.Delete(ctx, envCtx); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to delete external resource %s: %w", resourceHandler.Name(), err)
+		}
+
+		exists, err := resourceHandler.GetCurrentState(ctx, envCtx)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to get current state of external resource %s: %w", resourceHandler.Name(), err)
+		}
+
+		if exists != nil {
+			pendingDpResourcesDeletion = true
+		}
```

Comment:
Shall we use the improved finalizing approach as a deployment controller?

https://github.com/openchoreo/openchoreo/blob/main/internal/controller/deployment/controller_finalize.go#L84-L107

Reply:
+1 fixed..